### ISSUE TITLE
Broken faces

### ISSUE THREAD

**Issue by M-Ro** [jdolan/quetoo#669]
Certain faces often appear broken after map compile. This can be seen on Edge and a few other maps. I've isolated and attached a testcase boxmap below.

The polygon that is "broken" isn't consistent and often flips around to the other side of the torch sconce when recompiling the map.

Fiddling with texture surface params (texscale, rotation, or vert/horiz shift) does not appear to have any effect.

Moving the brush in worldspace seems to fix the face.

![quetoo014](https://user-images.githubusercontent.com/433127/153462921-3bbc9c48-a31f-40b4-ba54-74169963b5b4.png)

[testcase.zip](https://github.com/jdolan/quetoo/files/8043007/testcase.zip)  (.bsp & .map)

**Comment by jdolan** [comment]
Well, shit. This compiles / looks correct on my system. How should I jiggle the torch to repro?

**Comment by jdolan** [comment]
@M-Ro Care to retest this with latest? I'm simply skipping over the malformed triangles when calculating tangents. It could have just been in my head, but I swear, specular actually looked noticeably better on Torn Glory with this change, too.

**Comment by M-Ro** [comment]
I can't see any difference in regards to the white/black faces. Has something changed with specular? I'm not sure if I'm just noticing it now since you've mentioned it, but it seems much higher now, almost being blown out to the point of posterization in places.

![quetoo023](https://user-images.githubusercontent.com/433127/153679566-c6080c13-4d0a-4c9a-a7c4-a3da308498ad.png)
![quetoo024](https://user-images.githubusercontent.com/433127/153679574-6633e8ad-bf64-4db6-a8d3-0f689142680e.png)

**Comment by Panjoo** [comment]
> They are still present on Edge after recompiling the map.

Yes, I've noticed these 'holes' too and I think it has to do with the intersecting clipbrushes on those rocks. I've never seen clipbrushes create visible holes before so I'd say this is a bug. As a test, delete the clipbrushes there and recompile.

**Comment by jdolan** [comment]
The holes are a different issue. Holes are welding errors. White triangles are bad tangent vectors. I can see both in the Edge screenshots above, but.. just to clarify, we are talking about two separate problems.

Deleting clip brushes, or changing any brushes near this area, would quite conceivably "fix" the problem by changing the BSP tree, i.e. the order in which the brushes are split up by all of the planes in the map. It _could_ be something to do with clip brushes, but it could also just be that by modifying the brushes in that area, you end up with a different permutation of BSP splits -> leafs -> portals, etc.. that does not create this problem.

Basically, I think we can (did?) fix the majority of these. I'm not sure it's practical to spend much more time attempting to resolve _all_ of these issues, when they are rare and usually easily worked around. But I'm open to what others think..

**Comment by jdolan-chwy** [comment]
Planes, yes. Verts, no. Clip brushes do not generate portals / faces, and so there are no windings of verts for them at all. Clip brushes also _should_ not generate BSP splits. If you're seeing that, and you're sure there are no structural brushes using the same planes, then that would be a bug for sure. Do you have screenshots that show what you suspect is wrong here?

**Comment by Panjoo** [comment]
I'm combining screenshots that are taken in wireframe mode with normal screenshots, so you can see the lines in texture mode.

These are definitely cut lines from the clips:
![crazyclippy2](https://user-images.githubusercontent.com/10857572/153887095-8917aa8e-ec6e-45a4-b2d8-40ab7993e28b.jpg)
![crazyclippy1](https://user-images.githubusercontent.com/10857572/153888693-fc1f9933-63cc-4ca2-8765-99247946b54d.jpg)

**Comment by jdolan** [comment]
Nice. The missing sliver on the floor is fixed with this change, too, so.. hopefully I can get it merged down soon!

**Comment by Paril** [comment]
Under every bug is two other bugs. This is how programming works, sadly.

**Comment by Panjoo** [comment]
I'm aware, I've once heard someone say about making a videogame that it is like making a movie while at the same time inventing the camera. But because of the way this works it is also taking me a lot longer to finish up these maps. So plz bear with me. :P

**Comment by jdolan** [comment]
So the fix for clip brushes carving up the world was merged yesterday. Between that fix, and the bad tangent vectors fixes, do we want to consider this issue closed? There are still rare welding issues and bad tangents. I'm happy to leave this open but remove it from the 1.0 milestone...

**Comment by jdolan** [comment]
The one bad tangent I still saw on Torn Glory was fixed by just wiggling the vertex in Radiant and putting it right back to where it was 🤷  I think we've fixed these issues. Closing it out!
